localstack/utils/docker_utils.py

```diff
@@ -98,22 +98,50 @@ def add(self, port, mapped=None, protocol="tcp"):
         mapped = mapped or port
         if isinstance(port, list):
             for i in range(port[1] - port[0] + 1):
-                self.add(port[0] + i, mapped[0] + i)
+                if isinstance(mapped, list):
+                    self.add(port[0] + i, mapped[0] + i)
+                else:
+                    self.add(port[0] + i, mapped)
             return
         if port is None or int(port) <= 0:
             raise Exception("Unable to add mapping for invalid port: %s" % port)
         if self.contains(port):
+            print("port %s already in mappings" % (port))
             return
+        bisected_host_port = None
         for from_range, to_range in self.mappings.items():
             if not self.in_expanded_range(port, from_range):
                 continue
             if not self.in_expanded_range(mapped, to_range):
                 continue
-            self.expand_range(port, from_range)
-            self.expand_range(mapped, to_range)
+            from_range_len = from_range[1] - from_range[0]
+            to_range_len = to_range[1] - to_range[0]
+            is_uniform = from_range_len == to_range_len
+            if is_uniform:
+                self.expand_range(port, from_range)
+                self.expand_range(mapped, to_range)
+            else:
+                if not self.in_range(mapped, to_range):
+                    continue
+                # extending a 1 to 1 mapping to be many to 1
```

Comment:
Is there a test for this behavior?
Like:
```
ports.add(5000:5000)
ports.add(5001:5000)
assert ports.to_str() = "-p 5000-5001:5000"
```
or something like that? Or do I miss the point of this lines?

Reply:
@dfangl this case actually occurs when there is an adjacent port mapping to a many to one.  In this case it is in the extended range of the current mapping, but because a one to many map cannot be consolidated we continue the iteration.  This is covered by this case in the unit tests:
```
        port_mappings.add([7000, 7002], 7000)
        port_mappings.add(6999)
```

The example you've listed above actually is covered by adding any many to one mapping to the port mappings.  Because of the implementation where we split up a many to one and add each map recursively, the above example is actually equivalent to:

```
ports.add([5000, 5001], 5000)
```

In these cases we would not hit the if block above. Instead on each host-> container recurse we would simply hit this block each time we add a new mapping:
```
elif from_range_len == 1:
    self.expand_range(port, from_range)
```

Reply:
This last block was actually the one I was talking about, should have put my comment on a range instead of a single line (the comment).
I agree the existing tests should cover that behavior.